libopenage/datastructure/concurrent_queue.h

```diff
@@ -47,25 +46,48 @@ class ConcurrentQueue {
 		return this->queue.front();
 	}
 
-	/** Removes the front item of the queue and returns it. */
-	T &pop() {
-		std::unique_lock<mutex_t> lock{this->mutex};
-		while (this->queue.empty()) {
-			this->elements_available.wait(lock);
-		}
-		auto &item = this->queue.front();
+	/** Copies the front item in the queue and removes it from the queue. */
+	template<typename... None, typename U = T>
+	T pop([[maybe_unused]] typename std::enable_if_t<!std::is_move_constructible_v<U> and std::is_copy_constructible_v<U>>* t = NULL) {
+		static_assert(sizeof...(None) == 0, "User-specified template arguments are prohibited.");
+		std::scoped_lock lock{this->mutex};
+		T ret = this->front();
+		this->queue.pop();
+
+		// Explicitly call the copy constructor since T cannot be move-constructed
+		return T(ret);
```

Comment:
if you write `return ret`, c++ will select the best possible return way automatically (move or copy), we don't need to explicitly copy.

Reply:
`return ret` does not work on GCC 7.5.0 . Will try if it works for a later version.

Reply:
I tried to build with GCC 10.1.0, `return ret;` works here but there are several compile errors in other files.

Reply:
I tried this in godbolt now: https://godbolt.org/z/7_42Uy and the C++-Standard makes the difference. with `std=c++17` it works accross all versions, apparently (the copy-selection when move was deleted). `std=c++11` and `c++14` fail with use of deleted function.

since the buildsystem uses 17, i wonder why it fails for you :)

Reply:
Which version did you try on? On godbolt it also fails with GCC 7.5.0

Reply:
Only with gcc >7... and with =7 it fails. Oh dear. I think we can leave the explicit copy then, but with a `TODO` to change it once we raise our minimum compiler version.